sdk/eventhub/azure-eventhubs/azure/eventhub/_eventprocessor/ownership_manager.py

```diff
@@ -0,0 +1,142 @@
+                "Wrong partition id:{}. The eventhub has partitions: {}.".
+                    format(self.partition_id, self.cached_parition_ids))
+
+        if self.partition_manager is None:
+            return self.cached_parition_ids
+
+        ownership_list = self.partition_manager.list_ownership(
+            self.fully_qualified_namespace, self.eventhub_name, self.consumer_group_name
+        )
+        to_claim = self._balance_ownership(ownership_list, self.cached_parition_ids)
+        claimed_list = self.partition_manager.claim_ownership(to_claim) if to_claim else None
+        return [x["partition_id"] for x in claimed_list]
+
+    def _retrieve_partition_ids(self):
+        """List all partition ids of the event hub that the EventProcessor is working on.
+        """
+        self.cached_parition_ids = self.eventhub_client.get_partition_ids()
+
+    def _balance_ownership(self, ownership_list, all_partition_ids):
+        """Balances and claims ownership of partitions for this EventProcessor.
+        """
+
+        now = time.time()
+        ownership_dict = {x["partition_id"]: x for x in ownership_list}  # put the list to dict for fast lookup
+        not_owned_partition_ids = [pid for pid in all_partition_ids if pid not in ownership_dict]
+        timed_out_partitions = [x for x in ownership_list
+                                if x["last_modified_time"] + self.ownership_timeout < now]
+        if self._initializing:  # greedily claim all available partitions when an EventProcessor is started.
+            to_claim = timed_out_partitions
+            for to_claim_item in to_claim:
+                to_claim_item["owner_id"] = self.owner_id
+            for pid in not_owned_partition_ids:
+                to_claim.append(
+                    {
+                        "fully_qualified_namespace": self.fully_qualified_namespace,
+                        "partition_id": pid,
+                        "eventhub_name": self.eventhub_name,
+                        "consumer_group_name": self.consumer_group_name,
+                        "owner_id": self.owner_id
+                    }
+                )
+            self._initializing = False
+            if to_claim:  # if no expired or unclaimed partitions, go ahead with balancing
+                return to_claim
+
+        timed_out_partition_ids = [ownership["partition_id"] for ownership in timed_out_partitions]
+        claimable_partition_ids = not_owned_partition_ids + timed_out_partition_ids
+
+        active_ownership = [ownership for ownership in ownership_list
+                            if ownership["last_modified_time"] + self.ownership_timeout >= now]
+        active_ownership_by_owner = defaultdict(list)
+        for ownership in active_ownership:
+            active_ownership_by_owner[ownership["owner_id"]].append(ownership)
+        active_ownership_self = active_ownership_by_owner[self.owner_id]
+
+        # calculate expected count per owner
```

Comment:
Could this be moved into a separate function? 
Would make this method cleaner and unittesting easier :)

Reply:
sure thing, we'll considering moving in the next release.